src/Dictionaries/DirectDictionary.cpp

```diff
@@ -203,12 +211,15 @@ ColumnUInt8::Ptr DirectDictionary<dictionary_key_type>::hasKeys(
             block_keys_extractor.rollbackCurrentKey();
         }
 
+        keys_found += block_keys_size;
```

Comment:
If source will have duplicates this probably is wrong. Consider to move it up, and check for if ```(!result_data[result_data_found_index]) keys_found += 1;```, before ```result_data[result_data_found_index] = true;```